tests/unit/bokeh/models/test_sources.py

```diff
@@ -785,6 +785,53 @@ def test_data_column_lengths(self) -> None:
         assert len(warns) == 1
         assert str(warns[0].message) == "ColumnDataSource's columns must be of the same length. Current lengths: ('a', 3), ('b', 2)"
 
+class TestDataTable:
+    def test_from_data_with_dataframe(self):
+        df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
+        table = DataTable.from_data(df)
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) - {"index"} == {'A', 'B'}
+
+    def test_from_data_with_dict(self):
+        data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
+        table = DataTable.from_data(data)
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'B'}
+
+    def test_from_data_with_columndatasource(self):
+        data = ColumnDataSource({'A': [1, 2, 3], 'B': [4, 5, 6]})
+        table = DataTable.from_data(data)
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'B'}
+
+    def test_from_data_with_columns(self):
+        df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
+        table = DataTable.from_data(df, columns=['A', 'B'])
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'B'}
+
+        source = ColumnDataSource({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
+        table = DataTable.from_data(source, columns=['A', 'C'])
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'C'}
```

Comment:
This test could also check the `dict` input case

Reply:
In case you wanted to be a little fancy, this could be parametrized, something like this (untested):
```python
@pytest.mark.parametrize("typ", (dict, pd.DataFrame, ColumnDataSource))
def test_from_data_with_columns(self, typ):
    data = typ({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
    table = DataTable.from_data(data, columns=['A', 'B'])
    assert isinstance(table, DataTable)
    assert set(table.source.data) == {'A', 'B'
```

Reply:
Nice!